### ISSUE TITLE
February 2021 Release

### ISSUE THREAD

**Issue by toasted-nutbread** [FooSoft/yomichan#1472]
https://github.com/FooSoft/yomichan/releases/tag/21.2.28.1

@FooSoft 

### General
* [x] Create [release tag](https://github.com/FooSoft/yomichan/tree/21.2.28.1).
* [x] Create [GitHub release](https://github.com/FooSoft/yomichan/releases/tag/21.2.28.1).

### Testing
* [x] Update Chrome testing release.
* [x] Update Firefox testing release.
* [x] Update Firefox [updates.json](https://github.com/FooSoft/foosoft.github.io/blob/master/projects/yomichan/dl/updates.json) and [verify](https://foosoft.net/projects/yomichan/dl/updates.json).
* [x] Upload testing builds to the [GitHub release](https://github.com/FooSoft/yomichan/releases/tag/21.2.28.1).

### Stable
* [x] Update Chrome stable release.
* [x] Update Firefox stable release.
* [x] Upload stable builds to the [GitHub release](https://github.com/FooSoft/yomichan/releases/tag/21.2.28.1).

### Issues to address before stable
* [x] #1476
* [x] #1482
* [x] #1483
* [x] #1486
* [x] #1489
* [x] #1491, #1492
* [x] #1493
* [x] #1495
* [x] #1496
* [x] #1507

**Comment by Thermospore** [comment]
Nice! Tempted to switch over to the testing release. Is there a recommended procedure for this?

**Comment by toasted-nutbread** [comment]
Just install it the same as the normal release, then install dictionaries and copy over settings (works best in that order). Then you can just disable or uninstall the stable version so there aren't two active versions.

**Comment by FooSoft** [comment]
Testing release updated!

**Comment by Thermospore** [comment]
What are `enableWanakana` and `selectedParser`?
![image](https://user-images.githubusercontent.com/35350819/109762261-d4309380-7b9c-11eb-8eb8-d10d94341428.png)
![image](https://user-images.githubusercontent.com/35350819/109762659-718bc780-7b9d-11eb-9190-cc6391fe27ac.png)

Just installed testing release and hand copied the settings (bit paranoid I guess haha). Looking at the diff of the settings export

**Comment by Thermospore** [comment]
Not new, but is there a purpose to selecting a freq dict as a secondary search dict?
![image](https://user-images.githubusercontent.com/35350819/109763683-e7dcf980-7b9e-11eb-85ca-6bc196a06c6d.png)

**Comment by Thermospore** [comment]
I like the new color of that name classification tag
![image](https://user-images.githubusercontent.com/35350819/109765247-1065f300-7ba1-11eb-968c-8ecd0e05e21e.png)　![image](https://user-images.githubusercontent.com/35350819/109765270-165bd400-7ba1-11eb-9fc5-bacc9f3ca695.png)

very nice to be able to set freq dict priority!

interesting, I'm guessing the old tag text spacing was set with only lowercase latin characters in mind. the new style looks nice
![image](https://user-images.githubusercontent.com/35350819/109765836-d6e1b780-7ba1-11eb-8df3-5f58403799b1.png)　![image](https://user-images.githubusercontent.com/35350819/109765857-df39f280-7ba1-11eb-98a3-bd8d61ab75fc.png)

The tags are a bit taller now though, I preferred the previous height

**Comment by Thermospore** [comment]
Uh oh here is a significant issue, things are missing from the search :( Maybe related to `enableWanakana` and `selectedParser`?
![image](https://user-images.githubusercontent.com/35350819/109768437-5fae2280-7ba5-11eb-89dc-a056965b2f15.png)

something weird going on, there is text missing from the sentence display thing too
![image](https://user-images.githubusercontent.com/35350819/109769257-6be6af80-7ba6-11eb-95c5-23aa55594623.png)

![image](https://user-images.githubusercontent.com/35350819/109769668-fc24f480-7ba6-11eb-871c-0395913183ac.png)
![image](https://user-images.githubusercontent.com/35350819/109769972-6b024d80-7ba7-11eb-8ac2-9e42a364a554.png)

### files:
[old version settings.zip](https://github.com/FooSoft/yomichan/files/6074286/old.version.settings.zip) (21.1.31.1)
[new version settings.zip](https://github.com/FooSoft/yomichan/files/6074287/new.version.settings.zip) (21.2.28.0)
[[Monolingual] 三省堂 大辞林 第三版.zip](https://github.com/FooSoft/yomichan/files/6074308/Monolingual.zip)
[新明解国語辞典第五版v3 [yoga].zip](https://github.com/FooSoft/yomichan/files/6074311/v3.yoga.zip)

**Comment by toasted-nutbread** [comment]
> Just installed testing release and hand copied the settings (bit paranoid I guess haha). Looking at the diff of the settings export

`enableWanakana` corresponds to the _Automatic kana conversion_ toggle on the search page.
`selectedParser` corresponds to the different parsing algorithms for parsing the block of text below the search box. `"scan"` and `null` should effectively be the same, but I should probably unify that at some point. A different parser can be used if MeCab is installed.

**Comment by toasted-nutbread** [comment]
> Not new, but is there a purpose to selecting a freq dict as a secondary search dict?

No, it's only used to find additional definitions in merged mode, although the benefit of having this as a separate option is debatable.

**Comment by toasted-nutbread** [comment]
> something weird going on, there is text missing from the sentence display thing too

Fixed in #1483.

**Comment by toasted-nutbread** [comment]
> The tags are a bit taller now though, I preferred the previous height

It's a difference of 1.5 pixels (18.5px => 20px), which makes the split tags able to hold content with a standard line height (20px).

**Comment by toasted-nutbread** [comment]
> Uh oh here is a significant issue, things are missing from the search

The dictionary priorities would seem to be the issue here. The results are still there, just lower.

**Comment by Thermospore** [comment]
Thanks, I appreciate the responses :v:

> The dictionary priorities would seem to be the issue here. The results are still there, just lower.

Oh hey, they are. My dictionary priorities are in the same order, so I'm guessing the sort algo was changed?

The new way is kind of a game breaker, it was easiest to show in a video (sorry if I'm jumping the gun here)
https://youtu.be/pU7CaU-yvjE

Could that sort algo be reverted? or add an option to use the old one?

**Comment by Thermospore** [comment]
The reading-conscious freq dictionaries actually are missing sometimes, as far as I can tell
![image](https://user-images.githubusercontent.com/35350819/110054178-1a096b00-7d07-11eb-9477-a7433fce962e.png)
![image](https://user-images.githubusercontent.com/35350819/110054186-1ece1f00-7d07-11eb-9c56-b1688ef6ac39.png)

**Comment by toasted-nutbread** [comment]
> Could that sort algo be reverted? or add an option to use the old one?

The sorting algorithm actually wasn't changed, it was just enabled for situations where it previously wasn't. The issue is likely that the dictionary priority is used before all other sorting conditions, which make some shorter words appear before longer words.

https://github.com/FooSoft/yomichan/blob/ca4d5a2ac19d1e97025b579157568fb1e9fa0701/ext/js/language/translator.js#L1297-L1326

So the comparison of `dictionaryOrder` should probably come after the `source.length` and `reasons.length` comparisons.

**Comment by toasted-nutbread** [comment]
#1486 fixes the tag issues.

Unrelated, but #1488 is a layout bug with the new frequency tag `inline-list` style on Firefox. Fixed with #1489.

**Comment by Thermospore** [comment]
> The sorting algorithm actually wasn't changed, [etc]

Ahh interesting. Thanks for the explanation

> #1486 fixes the tag issues.

Nice. Yea I notice there tends to be inconsistency with how dictionaries store kana only words, which leads to problems. eg the grouping breaking
<details>
  <summary>click to expand; somewhat nsfw</summary>
  
  (this screenshot is from 21.1.31.1)
  ![image](https://user-images.githubusercontent.com/35350819/110196306-1ef61980-7df0-11eb-82c6-25f0d88999b4.png)
  
</details>

**Comment by Thermospore** [comment]
Noticed a new bug: when you copy paste out of yomichan into anki, the newlines are missing now. They are still there when you paste into the github comment box though

![image](https://user-images.githubusercontent.com/35350819/110196786-629e5280-7df3-11eb-9b59-1e98c4259c30.png)
anki html:
```
21.1.31.1<br>しん えん [0]【深▼淵】<br>①深いふち。<br>②奥深さや限界が底知れないことのたとえ。「欲望の━」「悲しみの━」<br><br>21.2.28.0<br>しん えん [0]【深▼淵】 ①深いふち。 ②奥深さや限界が底知れないことのたとえ。「欲望の━」「悲しみの━」<br>
```

**Comment by toasted-nutbread** [comment]
> Nice. Yea I notice there tends to be inconsistency with how dictionaries store kana only words, which leads to problems. eg the grouping breaking

This is an issue that's on the radar, but it will take a bit of work to address all the places in the code that this edge case is handled.

> Noticed a new bug: when you copy paste out of yomichan into anki, the newlines are missing now. They are still there when you paste into the github comment box though

This is probably because I removed `<br>` elements in favor of CSS `white-space:pre-line;`, but looks like I'll have to revert that.

**Comment by Thermospore** [comment]
Would #1497 cause any issues for entries with a blank reading but kanji in the expression? For example, this copy of 大辞林 leaves the reading field blank for ことわざ
![image](https://user-images.githubusercontent.com/35350819/110224612-90ce7180-7e89-11eb-8ddc-cd6d8652aaf4.png)

**Comment by toasted-nutbread** [comment]
It shouldn't; the handling of empty readings is treated as a special case in several locations throughout the code, and it just assigns `reading = expression`. The change just makes it so that this assignment occurs only once instead of several times at different locations.

**Comment by toasted-nutbread** [comment]
@FooSoft I've made a new release which addresses the changes listed in the opening post. I've also updated the OP to point to the new version, so the checkboxes are unticked (since only the testing release was made).

https://github.com/FooSoft/yomichan/releases/tag/21.2.28.1

This one should probably be safe enough to put to stable, but you can do a testing release for a few more days if you (or anyone else) likes.

**Comment by Thermospore** [comment]
np, thanks for putting up with me 🙂

**Comment by FooSoft** [comment]
Released to both test and stable 🎉

**Comment by Thermospore** [comment]
Think I found some bugs with the frequency display formatting and the entry sorting: https://youtu.be/DsYlLvgeVoc

settings file: [yomichan-settings-2021-03-10-06-51-58.zip](https://github.com/FooSoft/yomichan/files/6113832/yomichan-settings-2021-03-10-06-51-58.zip)

```
Extension version: Yomichan (development build) 21.2.28.1
Platform: Windows
Browser: Chrome
Language: ja
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.190 Safari/537.36
AnkiConnect version: Unknown (Anki not running or connected)
Installed dictionaries: JMdict, 新明解国語辞典 第五版, 大辞林３, Weblio 古語辞典, KANJIDIC (English), JMnedict, B短, B長, IC, AJ, W, 明鏡国語辞典, 故事ことわざの辞典, RikaiCake [lower], 実用日本語表現辞典, 漢字源, RikaiCake [upper]
```

**Comment by toasted-nutbread** [comment]
The tag issue is due to a fixed height for tags, which should have been a minimum. As a temporary workaround, use this CSS:

```css
.tag {
    height: auto;
    min-height: var(--tag-height);
}
```

The dictionary order is due to another issue with exact kana/term matches. This is partially stemming from the confusing data types used to represent a dictionary entry, which I am working on resolving, but I'll try to get a temporary fix for the issue in a patch.

**Comment by toasted-nutbread** [comment]
## [Patch release 21.2.28.2](https://github.com/FooSoft/yomichan/releases/tag/21.2.28.2)

@FooSoft 

### General
* [ ] Create [release tag](https://github.com/FooSoft/yomichan/tree/21.2.28.2).
* [ ] Create [GitHub release](https://github.com/FooSoft/yomichan/releases/tag/21.2.28.2).

### Testing
* [ ] Update Chrome testing release.
* [ ] Update Firefox testing release.
* [ ] Update Firefox [updates.json](https://github.com/FooSoft/foosoft.github.io/blob/master/projects/yomichan/dl/updates.json) and [verify](https://foosoft.net/projects/yomichan/dl/updates.json).
* [ ] Upload testing builds to the [GitHub release](https://github.com/FooSoft/yomichan/releases/tag/21.2.28.2).

### Stable
* [ ] Update Chrome stable release.
* [ ] Update Firefox stable release.
* [ ] Upload stable builds to the [GitHub release](https://github.com/FooSoft/yomichan/releases/tag/21.2.28.2).

**Comment by FooSoft** [comment]
Shipped!

**Comment by toasted-nutbread** [comment]
Thanks as always

**Comment by Thermospore** [comment]
Everything working great for me so far. Here is one more bug I noticed with the query parser tho:
![無題](https://user-images.githubusercontent.com/35350819/110885874-17bd8880-8294-11eb-9d45-9facacee2d3a.png)
![スクリーンショット 2021-03-11 175818](https://user-images.githubusercontent.com/35350819/110885876-18561f00-8294-11eb-8489-190a75619b08.png)　![スクリーンショット 2021-03-11 180025](https://user-images.githubusercontent.com/35350819/110885878-18eeb580-8294-11eb-9a9c-887e97feecf6.png)　![スクリーンショット 2021-03-11 180040](https://user-images.githubusercontent.com/35350819/110885879-18eeb580-8294-11eb-8d11-f291ba12e436.png)
I'm not sure if it's kana specific or if it could happen with other translations

doesn't happen when `Convert hiragana to katakana` and `Convert katakana to hiragana` are disabled
